client/tso_service_discovery.go

```diff
@@ -251,69 +416,129 @@ func (c *tsoServiceDiscovery) SetTSOGlobalServAddrUpdatedCallback(callback tsoGl
 
 // getPrimaryAddr returns the primary address.
 func (c *tsoServiceDiscovery) getPrimaryAddr() string {
-	primaryAddr := c.primary.Load()
-	if primaryAddr == nil {
-		return ""
-	}
-	return primaryAddr.(string)
+	c.keyspaceGroupSD.RLock()
+	defer c.keyspaceGroupSD.RUnlock()
+	return c.keyspaceGroupSD.primaryAddr
 }
 
 // getSecondaryAddrs returns the secondary addresses.
 func (c *tsoServiceDiscovery) getSecondaryAddrs() []string {
-	secondaryAddrs := c.secondaries.Load()
-	if secondaryAddrs == nil {
-		return []string{}
-	}
-	return secondaryAddrs.([]string)
+	c.keyspaceGroupSD.RLock()
+	defer c.keyspaceGroupSD.RUnlock()
+	return c.keyspaceGroupSD.secondaryAddrs
 }
 
-func (c *tsoServiceDiscovery) switchPrimary(addrs []string) error {
-	// FIXME: How to safely compare primary urls? For now, only allows one client url.
-	addr := addrs[0]
-	oldPrimary := c.getPrimaryAddr()
-	if addr == oldPrimary {
-		return nil
-	}
-
-	if _, err := c.GetOrCreateGRPCConn(addr); err != nil {
-		log.Warn("[tso] failed to connect primary", zap.String("primary", addr), errs.ZapError(err))
-		return err
-	}
-	// Set PD primary and Global TSO Allocator (which is also the PD primary)
-	c.primary.Store(addr)
+func (c *tsoServiceDiscovery) afterPrimarySwitched(oldPrimary, newPrimary string) error {
 	// Run callbacks
 	if c.globalAllocPrimariesUpdatedCb != nil {
-		if err := c.globalAllocPrimariesUpdatedCb(addr); err != nil {
+		if err := c.globalAllocPrimariesUpdatedCb(newPrimary); err != nil {
 			return err
 		}
 	}
-	log.Info("[tso] switch primary", zap.String("new-primary", addr), zap.String("old-primary", oldPrimary))
+	log.Info("[tso] switch primary",
+		zap.String("new-primary", newPrimary),
+		zap.String("old-primary", oldPrimary))
 	return nil
 }
 
-func (c *tsoServiceDiscovery) updateMember() error {
-	resp, err := c.metacli.Get(c.ctx, []byte(c.primaryKey))
+func (c *tsoServiceDiscovery) updateMember() (err error) {
+	// The keyspace membership or the primary serving address of the keyspace group, to which this
+	// keyspace belongs, might have been changed. We need to query tso servers to get the latest info.
+	// If there are known healthy tso servers serving this keyspace, we start with them; otherwise,
+	// we start with any healthy tso server.
+	var keyspaceGroup *tsopb.KeyspaceGroup
+	tsoServerAddr := c.keyspaceGroupSD.getServer()
+	if len(tsoServerAddr) > 0 {
+		keyspaceGroup, err = c.findGroupByKeyspaceID(c.keyspaceID, tsoServerAddr, updateMemberTimeout)
```

Comment:
Is there any compatibility issue?

Reply:
Yes, I added a new rpc FindGroupByKeyspaceID in tso service. If we upgrade client ahead of tso server side, there will be compatibility issue. We can check with serviceless service team to see if we need to handle this case. If yes, we can check if the error contains "unimplemented" error and fallback to metastorage client to retrieve the primary address of the default keyspace group, and it can be done in a separate pr.